elastic/kibana
# [Embeddable] Clientside migration system

Closes https://github.com/elastic/kibana/issues/158677
Closes https://github.com/elastic/kibana/issues/161816

# Summary
This PR changes the versioning scheme used by Dashboard Panels and by value Embeddables, and introduces a new clientside system that can migrate Embeddable Inputs to their latest versions.

# Embeddable Migration System
> **Note**
This PR does not affect the existing serverside Embeddable migrations in any way - it only adds a secondary system on top that can be used without adding an additional saved object migration to Dashboard. Older Embeddable migrations will continue to run against older inputs.

### How are versions stored?
After this PR Embeddable versions and the migrations between are no longer tied to the Kibana version. Instead, each Embeddable type maintains its own versioning information. This is stored as follows:

- the **Latest version** is stored in the Embeddable Factory. If a latest version is not present, no migrations will be run against this embeddable type.
- the **Current version** is stored in the Embeddable's Input. 

### When do migrations get run?
The migrations get run as part of the `embeddableFactory.create` process. This means that every Embeddable that gets rendered by any application will receive migrations. Persisting an Embeddable in a saved object no longer means worrying about keeping them up to date or writing any migrations. Just save the input as-is, and pass it to an Embeddable Renderer, which will do the rest.

On Dashboards... there are some extra steps.

# Dashboard Embeddable Migrations
Usually the process of just rendering an Embeddable and having the migrations happen naturally is enough. For Dashboard this is not the case because:
- Dashboards need to know if a migration has happened in order to prompt the user to save.
- Dashboards remember their `last saved state` to allow for resetting and diff checking. Resetting a Dashboard should not undo migrations!

For these reasons, Dashboards run a function clientside on saved object load. This function migrates all panels and Controls on the Dashboard _before_ the Embeddables are even created. If this function detects any changes a save prompt will be shown

### Save Prompt
Saving back-propagates any changes from the migrations back to the saved object, and can make loading faster next time because migrations can be skipped. To encourage users to do this if changes have been detected and the Dashboard is in edit mode a prompt will be shown next to the `unsaved changes` badge. Here is the prompt in context:

<img width="1731" alt="Screenshot 2023-08-11 at 6 04 02 PM" src="https://github.com/elastic/kibana/assets/14276393/104eefcd-5b48-46ad-964c-4db714aebf08">


# Old Version Key
There is a version key in the Dashboard panel type which has historically stored the Kibana version the Dashboard was last saved on.

In this PR that version key was removed and version information was moved inside the Input key so that **all** Embeddables can be versioned, not just the ones stored in Dashboards. As part of that change, the two usages of the old version key needed to be updated to reflect that the older version key can be missing.

The two usages were:
1. To check if the panel was saved in a version before 7.3.0 in order to use a legacy version of the extract references function. This PR removes that legacy version of the extract method, because the migration that required it would never be run on a version after 8.0 because upgrading to 8.x requires a stop at 7.17 first.
2. To check if a panel injected via the URL is older than 7.3.0 in order to abort loading it. This has been updated to also check directly for the hallmarks of an older style panel, like the lack of a `gridData` key.

# How to test
`src/plugins/image_embeddable/public/image_embeddable/image_embeddable_factory.tsx`
``` typescript
public latestVersion = '1.0.1';
public migrations = {
  '1.0.1': (input: ImageEmbeddableInput) => {
    input.imageConfig.altText = `Neat migration! Previous alt text was "${input.imageConfig.altText}"`;
    return input;
  },
};
```

The new `save prompt` should show, but the `unsaved changes` badge should **not** show. Resetting the input, switching back and forth between view & edit mode and any action done on the Dashboard should not undo the migration, or run the migration again on already migrated input.

If the migration runs against the input more than once, you'll see the alt text become: `Neat migration! Previous alt text was "Neat migration! Previous alt text was "Whatever your alt text was originally""`.

# What is not covered?
This PR is a very basic and simple version of the clientside migration system. More functionality and features can be added to it as they are needed. Because of this, some pieces are missing:
- Right now, the Dashboard does not run clientside **Control Group Migrations**. This is because in order to do so, the Dashboard Saved Object would need to optionally store a Control Group version, which would require changes in the saved object schema, and the content management. Before we can implement any migrations for the Control Group itself we will need to make these changes. Fortunately, migrations are quite rare for the Control Group itself.
- **Drilldown Migrations** are not covered at the moment because in order for them to be implemented we would need to extend the Embeddable Enhancements system to be able to store version information independent from the actual Embeddable version. There are currently no migrations registered for drilldowns, but before we add one, we would need to ensure that a system for properly versioning them is in place.
- **Propagation from child Embeddables back to Container**
The function that runs on Dashboard saved object load means that most of the time, when the embeddable create function is called on Dashboards there are no more migrations to be run. If for any reason there are migrations to be run at this point, the embeddable inheritance structure means they get immediately undone because the Container's child input is considered the source of truth. This could happen when:
  - The Dashboard has state in the **session storage**. If a user has unsaved panels in their session storage, they refresh the Dashboard, and Kibana has updated, the Embeddable create method will detect if any new migrations have been registered and will run them, **but** the container will pick up the changes and immediately undo them.
  - There are panels from an older version injected via the **URL** from a snapshot link. Again the migrations will get run,  but will immediately get undone.

  Implementing this back-propagation on the existing Embeddables framework would require ugly workarounds and changes that would affect the stability of the Embeddable platform. Once we move to a less prescriptive inheritance structure, this will not be a problem.

### Checklist

Delete any items that are not applicable to this PR.
